engines/wintermute/base/base_keyboard_state.cpp

```diff
@@ -403,8 +408,222 @@ enum VKeyCodes {
+	case Common::KEYCODE_F1:
+		return kFtVkF1;
+	case Common::KEYCODE_F2:
+		return kFtVkF2;
+	case Common::KEYCODE_F3:
+		return kFtVkF3;
+	case Common::KEYCODE_F4:
+		return kFtVkF4;
+	case Common::KEYCODE_F5:
+		return kFtVkF5;
+	case Common::KEYCODE_F6:
+		return kFtVkF6;
+	case Common::KEYCODE_F7:
+		return kFtVkF7;
+	case Common::KEYCODE_F8:
+		return kFtVkF8;
+	case Common::KEYCODE_F9:
+		return kFtVkF9;
+	case Common::KEYCODE_F10:
+		return kFtVkF10;
+	case Common::KEYCODE_F11:
+		return kFtVkF11;
+	case Common::KEYCODE_F12:
+		return kFtVkF12;
+	case Common::KEYCODE_NUMLOCK:
+		return kFtVkNumLock;
+	case Common::KEYCODE_SCROLLOCK:
+		return kFtVkScroll;
+	default:
+		// check if any non-sticky keys were used, otherwise key is unknown to us
+		if ((event->kbd.flags & Common::KBD_NON_STICKY) == 0) {
+			warning("Key pressed is not recognized, ASCII returned (%d '%c').", event->kbd.keycode, event->kbd.keycode);
+		}
+		// return ASCII if no match, since it could be used for typing
+		return event->kbd.ascii;
+		break;
+	}
+}
+
+//////////////////////////////////////////////////////////////////////////
+Common::KeyCode vKeyToKeyCodeFoxtail(uint32 vkey) {
+	switch (vkey) {
+	case kFtVkBack:
+		return Common::KEYCODE_BACKSPACE;
+	case kFtVkTab:
+		return Common::KEYCODE_TAB;
+	case kFtVkClear:
+		return Common::KEYCODE_CLEAR;
+	case kFtVkReturn:
+		return Common::KEYCODE_RETURN;
+	case kFtVkPause:
+		return Common::KEYCODE_PAUSE;
+	case kFtVkCapital:
+		return Common::KEYCODE_CAPSLOCK;
+	case kFtVkEscape:
+		return Common::KEYCODE_ESCAPE;
+	case kFtVkSpace:
+		return Common::KEYCODE_SPACE;
+	case kFtVkPrior:
+		return Common::KEYCODE_PAGEUP;
+	case kFtVkNext:
+		return Common::KEYCODE_PAGEDOWN;
+	case kFtVkHome:
+		return Common::KEYCODE_HOME;
+	case kFtVkEnd:
+		return Common::KEYCODE_END;
+	case kFtVkLeft:
+		return Common::KEYCODE_LEFT;
+	case kFtVkRight:
+		return Common::KEYCODE_RIGHT;
+	case kFtVkUp:
+		return Common::KEYCODE_UP;
+	case kFtVkDown:
+		return Common::KEYCODE_DOWN;
+	case kFtVkPrint:
+		return Common::KEYCODE_PRINT;
+	case kFtVkInsert:
+		return Common::KEYCODE_INSERT;
+	case kFtVkDelete:
+		return Common::KEYCODE_DELETE;
+	case kFtVkF1:
+		return Common::KEYCODE_F1;
+	case kFtVkF2:
+		return Common::KEYCODE_F2;
+	case kFtVkF3:
+		return Common::KEYCODE_F3;
+	case kFtVkF4:
+		return Common::KEYCODE_F4;
+	case kFtVkF5:
+		return Common::KEYCODE_F5;
+	case kFtVkF6:
+		return Common::KEYCODE_F6;
+	case kFtVkF7:
+		return Common::KEYCODE_F7;
+	case kFtVkF8:
+		return Common::KEYCODE_F8;
+	case kFtVkF9:
+		return Common::KEYCODE_F9;
+	case kFtVkF10:
+		return Common::KEYCODE_F10;
+	case kFtVkF11:
+		return Common::KEYCODE_F11;
+	case kFtVkF12:
+		return Common::KEYCODE_F12;
+	case kFtVkNumLock:
+		return Common::KEYCODE_NUMLOCK;
+	case kFtVkScroll:
+		return Common::KEYCODE_SCROLLOCK;
+	default:
+		warning("Unknown VKEY: %d", vkey);
+		return (Common::KeyCode)(vkey < KEYSTATES_ARRAY_SIZE ? vkey : 0);
+		break;
+	}
+}
+#endif
+
```

Comment:
Overall, there is a lot of duplication here, and this is prone to errors (e.g. adding a case in one function, but not the other). Also, the Foxtail enums are only used here, so IMHO we don't really need them.

You could use a simple list, which would contain the ScummVM keycode and the Foxtail keycode in each entry, e.g:

    struct FoxTailKeyCodeMapping {
        uint32 foxTailKeyCode;
        Common::KeyCode keycode;
    };

        const FoxTailKeyCodeMapping foxTailKeyCodeMapping[] = {
                { 1073741898, Common::KEYCODE_HOME },
                { 1073741882, Common::KEYCODE_F1 }
        };

And then you could do something like:

uint32 keyCodeToFoxTailKeyCode(Common::Event* event) {
	for (int i = 0; i < ARRAYSIZE(foxTailKeyCodeMapping); i++) {
		if (foxTailKeyCodeMapping[i].keycode == event->kbd.keycode)
			return foxTailKeyCodeMapping[i].foxTailKeyCode;
	}

	return event->kbd.ascii;
}

Common::KeyCode foxTailKeyCodeToKeyCode(uint32 vkey) {
	for (int i = 0; i < ARRAYSIZE(foxTailKeyCodeMapping); i++) {
		if (foxTailKeyCodeMapping[i].foxTailKeyCode == vkey)
			return foxTailKeyCodeMapping[i].keycode;
	}

	return Common::KEYCODE_INVALID;
}

Reply:
Very good idea, thanks for pointing this!

While implementing this suggestion and running various tests I discovered, that difference in key mappings between FoxTail and usual games is not due to FoxTail's engine differences from WME Lite, but were made in WME Lite itself! WME 1.x is using "Virtual-Key Codes" mapping
and WME Lite and FoxTail are using "SDL_Keycode" mapping.

So, here is total refactoring of keycode handling.

Why nobody noticed this earlier? Because there are only 4 currently known series that are based on WME Lite forks (J.U.L.I.A. series, Reversion series, FoxTail and Securanote) and because most WME games, both 1.x-based and Lite-based are using Keyboard class only for handling Enter and Escape keys, which are the same in both mappings (Backspace, Space and Tab are also the same).